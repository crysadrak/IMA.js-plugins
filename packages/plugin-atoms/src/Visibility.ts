/* eslint-disable jsdoc/require-param */
/* eslint-disable jsdoc/require-returns */
import {
  ClientWindow,
  Dependencies,
  Dispatcher,
  RouterEvents,
} from '@ima/core';
// @ts-expect-error
import { Circle } from 'infinite-circle';

type NotifyPayload = {
  type: string;
};

type NotifyCallback = (payload: NotifyPayload) => void;

/**
 * Visibility helper.
 */
export default class Visibility {
  static $dependencies: Dependencies = ['$Window', '$Dispatcher'];

  private window: ClientWindow;
  private dispatcher: Dispatcher;
  private afterHandleRouteCalled = false;
  circle: Circle;

  constructor(window: ClientWindow, dispatcher: Dispatcher) {
    this.window = window;
    this.dispatcher = dispatcher;
    this.circle = this.createVisibilityCircle();
  }

  /**
   * Create visibility circle.
   */
  private createVisibilityCircle() {
    return new Circle({
      listen: (notify: any) => this.listenOnEvents(notify),
      unlisten: (notify: any) => this.unlistenOnEvents(notify),
    });
  }

  /**
   * Register handlers to visibility loop
   */
  register(
    read: (...args: any[]) => any,
    write: (...args: any[]) => any,
    meta = { visibilityInterval: 180 }
  ): number {
    const id = this.circle.register({
      read,
      write,
      meta: { interval: meta.visibilityInterval },
    });

    if (this.afterHandleRouteCalled) {
      this.notify({ id });
    }

    return id;
  }

  /**
   * Unregister handlers from visibility loop
   */
  unregister(id: number) {
    this.circle.unregister(id);
  }

  /**
   * It cut down calling the event handler for defined interval. The throttle
   * method use requestAnimationFrame function which is called during page
   * scrolling.
   */
  throttle(
    eventHandler: (...args: any[]) => any,
    interval: number,
    context: any
  ) {
    const window = this.window.getWindow();
    interval = interval || 0;
    let callTime = 0;
    let lastArguments: any = null;

    if (context) {
      eventHandler = eventHandler.bind(context);
    }

    if (!this.window.isClient()) {
      return eventHandler;
    }

    function suspendAction() {
      if (callTime <= Date.now() || !window.requestAnimationFrame) {
        callTime = 0;
        eventHandler(...lastArguments);
      } else {
        window.requestAnimationFrame(suspendAction);
      }
    }

    return function throttle(...rest: any[]) {
      lastArguments = rest;

      if (!callTime) {
        callTime = Date.now() + interval;
        suspendAction();
      }
    };
  }

  /**
   * The method add circle instance to be running in the next infinite loop.
   */
  notify(...rest: any[]) {
    this.circle.notify(...rest);
  }

  /**
   * The visibility helper start checking visibility of registered entries.
   */
  private listenOnEvents(notify: NotifyCallback) {
    this.dispatcher.listen(
      RouterEvents.BEFORE_HANDLE_ROUTE,
      this.beforeHandleRoute,
      this
    );
    this.dispatcher.listen(
      RouterEvents.AFTER_HANDLE_ROUTE,
      this.afterHandleRoute,
      this
    );
    this.window.bindEventListener(this.window.getWindow(), 'resize', notify);
    this.window.bindEventListener(this.window.getWindow(), 'scroll', notify);
  }

  /**
   * The visibility helper stop checking visibility of registered entries.
   */
  private unlistenOnEvents(notify: NotifyCallback) {
    this.dispatcher.unlisten(
      RouterEvents.BEFORE_HANDLE_ROUTE,
      this.beforeHandleRoute,
      this
    );
    this.dispatcher.unlisten(
      RouterEvents.AFTER_HANDLE_ROUTE,
      this.afterHandleRoute,
      this
    );
    this.window.unbindEventListener(this.window.getWindow(), 'resize', notify);
    this.window.unbindEventListener(this.window.getWindow(), 'scroll', notify);
  }

  /**
   * The method resets `afterHandleRoute` marker.
   */
  private beforeHandleRoute() {
    this.afterHandleRouteCalled = false;
  }

  /**
   * The method normalize routeInfo to {@notifyPayload}.
   */
  private afterHandleRoute(routeInfo: any) {
    this.afterHandleRouteCalled = true;

    const payload = Object.assign(
      { type: RouterEvents.AFTER_HANDLE_ROUTE },
      routeInfo
    );

    this.notify(payload);
  }
}
