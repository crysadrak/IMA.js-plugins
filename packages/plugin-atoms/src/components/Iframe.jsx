import { useComponentUtils } from '@ima/react-page-renderer';
import { memo, useEffect, useRef, useState } from 'react';

import { LAYOUT } from './layout';
import { Sizer } from './Sizer';

export const Iframe = memo(function IframeComponent({
  name,
  src,
  layout,
  className,
  width,
  height,
  loading,
  placeholder,
  ...rest
}) {
  const [visibleInViewport, setVisibleInViewport] = useState(false);
  const rootElement = useRef();
  const { $CssClasses, $UIComponentHelper } = useComponentUtils();

  useEffect(() => {
    let extendedPadding = Math.max(
      $UIComponentHelper.componentPositions.getWindowViewportRect().height / 2,
      500
    );

    const registeredVisibilityId = $UIComponentHelper.visibility.register(
      $UIComponentHelper.getVisibilityReader(rootElement.current, {
        useIntersectionObserver: true,
        extendedPadding,
        width,
        height,
      }),
      $UIComponentHelper.wrapVisibilityWriter((visibility, observer) => {
        if (visibility > 0) {
          observer && observer.disconnect();
          $UIComponentHelper.visibility.unregister(registeredVisibilityId);

          if (visibleInViewport === false) {
            setVisibleInViewport(true);
          }
        }
      })
    );

    return () => {
      $UIComponentHelper.visibility.unregister(registeredVisibilityId);
    };
  }, []);

  return (
    <div
      ref={rootElement}
      className={$CssClasses(
        {
          'atm-iframe': true,
          'atm-overflow': true,
          'atm-responsive': layout === LAYOUT.RESPONSIVE,
          'atm-fill': layout === LAYOUT.FILL,
          'atm-placeholder': !visibleInViewport,
        },
        className
      )}
      style={
        layout === LAYOUT.RESPONSIVE
          ? {}
          : {
              width: width ?? 'auto',
              height: height ?? 'auto',
            }
      }
    >
      {layout === LAYOUT.RESPONSIVE ? (
        <Sizer width={width} height={height} />
      ) : null}
      {visibleInViewport && (
        <iframe
          {...rest}
          src={$UIComponentHelper.sanitizeUrl(src)}
          name={name ?? src}
          width={width}
          height={height}
          loading={loading ?? 'lazy'}
          className={$CssClasses({
            'atm-fill': true,
            //'atm-placeholder': placeholder !== false,
          })}
        />
      )}
    </div>
  );
});
