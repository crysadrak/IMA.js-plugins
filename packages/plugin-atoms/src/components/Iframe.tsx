import { useComponentUtils } from '@ima/react-page-renderer';
import {
  memo,
  useEffect,
  useRef,
  useState,
  ComponentPropsWithoutRef,
} from 'react';

import { LAYOUT } from './layout';
import { Sizer } from './Sizer';

export const Iframe = memo(function IframeComponent({
  name,
  src,
  layout,
  className,
  width,
  height,
  loading = 'lazy',
  placeholder,
  ...rest
}: {
  name?: string;
  src: string;
  layout?: LAYOUT;
  className?: string;
  width: number;
  height: number;
  loading?: 'eager' | 'lazy';
  placeholder?: boolean;
} & Omit<ComponentPropsWithoutRef<'iframe'>, 'placeholder'>) {
  const [visibleInViewport, setVisibleInViewport] = useState(false);
  const rootElement = useRef<HTMLDivElement>(null);
  const { $CssClasses, $UIComponentHelper } = useComponentUtils();

  useEffect(() => {
    if (rootElement.current === null) {
      return;
    }

    const extendedPadding = Math.max(
      $UIComponentHelper.componentPositions.getWindowViewportRect().height / 2,
      500
    );

    const registeredVisibilityId = $UIComponentHelper.visibility.register(
      $UIComponentHelper.getVisibilityReader(rootElement.current, {
        useIntersectionObserver: true,
        extendedPadding,
        width,
        height,
      }),
      $UIComponentHelper.wrapVisibilityWriter((visibility, observer) => {
        if (visibility > 0) {
          observer && observer.disconnect();
          $UIComponentHelper.visibility.unregister(registeredVisibilityId);

          if (visibleInViewport === false) {
            setVisibleInViewport(true);
          }
        }
      })
    );

    return () => {
      $UIComponentHelper.visibility.unregister(registeredVisibilityId);
    };
  }, []);

  return (
    <div
      ref={rootElement}
      className={$CssClasses(
        {
          'atm-iframe': true,
          'atm-overflow': true,
          'atm-responsive': layout === LAYOUT.RESPONSIVE,
          'atm-fill': layout === LAYOUT.FILL,
          'atm-placeholder': !visibleInViewport,
        },
        className
      )}
      style={
        layout === LAYOUT.RESPONSIVE
          ? {}
          : {
              width: width ?? 'auto',
              height: height ?? 'auto',
            }
      }
    >
      {layout === LAYOUT.RESPONSIVE ? (
        <Sizer width={width} height={height} />
      ) : null}
      {visibleInViewport && (
        <iframe
          {...rest}
          src={$UIComponentHelper.sanitizeUrl(src)}
          name={name ?? src}
          width={width}
          height={height}
          loading={loading}
          className={$CssClasses({
            'atm-fill': true,
            //'atm-placeholder': placeholder !== false,
          })}
        />
      )}
    </div>
  );
});
