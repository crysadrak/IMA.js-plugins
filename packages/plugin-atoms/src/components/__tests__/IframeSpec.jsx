import { renderWithContext, waitFor } from '@ima/testing-library';

import Visibility from '../../Visibility';
import { Iframe } from '../Iframe';

describe('Iframe', () => {
  it('should render with noscript tag by default', async () => {
    const { container } = await renderWithContext(
      <Iframe src='example.html' />
    );

    expect(container.querySelector('noscript')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render without noscript tag if visible', async () => {
    const { app, container } = await renderWithContext(
      <Iframe src='example.html' />
    );

    // Mock getBoundingClientRect to simulate visibility
    container.firstChild.getBoundingClientRect = jest.fn(() => ({
      width: 100,
      height: 100,
      top: 0,
      left: 0,
      bottom: 100,
      right: 100,
    }));
    app.oc.get(Visibility).notify(); // Trigger visibility observer

    // The component needs to re-render to remove noscript tag, there is no generic way to wait for it,
    // so we need to wait for the noscript tag to disappear
    await waitFor(() =>
      expect(container.querySelector('noscript')).not.toBeInTheDocument()
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
