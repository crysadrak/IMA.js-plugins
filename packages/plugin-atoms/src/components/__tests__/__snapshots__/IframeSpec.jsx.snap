// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Iframe should render with noscript tag by default 1`] = `
<div
  class="atm-iframe atm-overflow atm-placeholder"
  style="width: auto; height: auto;"
>
  <noscript>
    &lt;iframe src="example.html" name="example.html" width="auto" height="auto" class="atm-fill" &gt;&lt;/iframe&gt;
  </noscript>
</div>
`;

exports[`Iframe should render without noscript tag if visible 1`] = `
<div
  class="atm-iframe atm-overflow"
  style="width: auto; height: auto;"
>
  <iframe
    class="atm-fill"
    name="example.html"
    src="example.html"
  />
</div>
`;
